### PR TITLE
BOAC-2711: Revamps admin page

### PR DESCRIPTION
**Pull request by lyttam**
https://jira.ets.berkeley.edu/jira/browse/BOAC-2844

The new table allows filtering by department, but that doesn't include not-real departments like Admin and Notes Only.  I'll submit another PR to expand filtering capabilities to include permissions and status.

### CODE REVIEW

**Comment by lyttam** [comment]
image

File: boac/api/user_controller.py [hunk 1]
```diff
@@ -108,13 +108,26 @@ def delete_university_dept_membership(university_dept_id, authorized_user_id):
...
+def authorized_users():
+    sort_users_by = util.get(request.args, 'sortUsersBy', None)
+    return tolerant_jsonify(_get_boa_users(sort_users_by))
+
+
 @app.route('/api/users/authorized_groups')
 @admin_required
 def authorized_user_groups():
     sort_users_by = util.get(request.args, 'sortUsersBy', None)
     return tolerant_jsonify(_get_boa_user_groups(sort_users_by))
```
Comment (johncrossman) [inline boac/api/user_controller.py]: @lyttam We can remove `/api/users/authorized_groups` endpoint, yes?

File: boac/api/user_controller.py [hunk 2]
```diff
@@ -108,13 +108,26 @@ def delete_university_dept_membership(university_dept_id, authorized_user_id):
     )
 
 
+@app.route('/api/users/authorized')
+@admin_required
+def authorized_users():
+    sort_users_by = util.get(request.args, 'sortUsersBy', None)
+    return tolerant_jsonify(_get_boa_users(sort_users_by))
```
Comment (johncrossman) [inline boac/api/user_controller.py]: @lyttam - No need to show deleted users on new Admin page? (I don't remember the requirement.)

File: src/store/modules/user.ts [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
 
 const state = {
   calnetUsersByCsid: {},
+  departments: undefined,
   preferences: {
     sortBy: 'last_name'
   },
   user: undefined,
-  userGroups: undefined
+  userGroups: undefined,
```
Comment (johncrossman) [inline src/store/modules/user.ts]: @lyttam - Can we get rid of `userGroups` (above) and `setUserGroups` (below)?

**Comment by johncrossman** [comment]
Looks really good @lyttam  Thank you!  I posted a few questions. I have one request: Would you mind adding a line of documentation in `authorized_user.py` to explain the `is_blocked` property? Its purpose would not be obvious if I were seeing it for the first time.

File: boac/api/user_controller.py (hunk 2 above)
Comment (lyttam) [reply boac/api/user_controller.py]: @johncrossman We will need to show them actually - I'll make that change.

File: src/store/modules/user.ts (hunk 3 above)
Comment (lyttam) [reply src/store/modules/user.ts]: Yes, I'll remove it.

File: boac/api/user_controller.py (hunk 1 above)
Comment (lyttam) [reply boac/api/user_controller.py]: @johncrossman Yes you're right, I'll remove it.

**Comment by lyttam** [comment]
@johncrossman I've addressed your points; this is ready for a second look.

File: boac/lib/http.py [hunk 4]
```diff
@@ -140,3 +140,8 @@ def response_with_csv_download(rows, filename_prefix, fieldnames=None):
     csv_writer.writeheader()
     csv_writer.writerows(rows)
     return response
+
+
+def _stringify_date(o):
+    if isinstance(o, datetime):
+        return o.__str__()
```
Comment (johncrossman) [inline boac/lib/http.py]: @lyttam What ends up in resulting JSON if `not isinstance(o, datetime)`?

Comment (lyttam) [reply boac/lib/http.py]: @johncrossman If `deleted_at` is populated with a `datetime`, I get a `TypeError: datetime.datetime is not JSON serializable`

Comment (johncrossman) [reply boac/lib/http.py]: @lyttam - Understood. The convention elsewhere (e.g, `advising_note.note_to_compatible_json()`)  has been `isoformat()` to stringify. Can we add that in `authorized_users_api_feed()`?  I'm nervous about overriding the default serializer.

File: src/components/admin/Users.vue [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+      const key = sortBy === 'name' ? 'lastName' : sortBy
+      let aValue = this.get(a, key);
+      let bValue = this.get(b, key);
+      return this.sortComparator(aValue, bValue);
+    }
+  }
+}
+</script>
+
+<style>
```
Comment (johncrossman) [inline src/components/admin/Users.vue]: @lyttam Should these styles be "scoped"?

File: src/components/admin/Users.vue [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    filterFields: ['depts'],
+    items: [],
+    perPage: 10,
+    rowCount: 0,
+    sortBy: 'lastName',
+    sortDesc: false
+  }),
+  created() {
+    this.items = this.cloneDeep(this.users);
+    this.rowCount = this.users ? this.users.length : 0;
```
Comment (johncrossman) [inline src/components/admin/Users.vue]: @lyttam - A minor concern: `this.size(this.users)" is alternative to line above.

File: src/components/admin/Users.vue (hunk 5 above)
Comment (lyttam) [reply src/components/admin/Users.vue]: @johncrossman Yes, except then the `class` attributes in the `fields` definition will be ignored.  Seems like a quirk to me but it is in the [documentation](https://bootstrap-vue.js.org/docs/components/table#field-definition-reference):  "class, thClass, tdClass etc. will not work with classes that are defined in scoped CSS, unless you are using VueLoader's Deep selector."

I could scope everything but those classes, if it seems worth it.

Comment (johncrossman) [reply src/components/admin/Users.vue]: Ah, thanks for clarifying. It's fine as is.

File: boac/api/util.py [hunk 7]
```diff
@@ -79,13 +80,18 @@ def authorized_users_api_feed(users, sort_by='lastName'):
         profile.update({
             'id': user.id,
             'isAdmin': user.is_admin,
+            'isBlocked': user.is_blocked,
+            'canAccessCanvasData': user.can_access_canvas_data,
+            'deletedAt': _isoformat(user.deleted_at),
```
Comment (lyttam) [inline boac/api/util.py]: @johncrossman I moved the stringification of the `datetime` to `authorized_users_api_feed()` as you suggested.

**Comment by johncrossman** [comment]
Thanks @lyttam  Waiting on Travis

**Pull request merged**